pkg/metrics/collector/pod_ip_metrics.go

```diff
@@ -318,26 +334,104 @@ func (c *podIPMetricsCollector) setupDirectoryWatcher(dir string) error {
+	if subnet.IP.To16() != nil && subnet.IP.To4() == nil {
+		c.assignedIPv6AddrCount += totalIP
+	} else if subnet.IP.To4() != nil {
+		c.assignedIPv4AddrCount += totalIP
+	}
+}
+
+func (c *podIPMetricsCollector) calculateAssignedIP(subnetFile string) error {
+	networkConfigList, err := libcni.ConfListFromFile(subnetFile)
+	if err != nil {
+		glog.Errorf("Error reading network config from file %s: %v", subnetFile, err)
+		return err
+	}
+
+	var ipamConfig *allocator.IPAMConfig
+	gkePluginFound := false
+	for _, plugin := range networkConfigList.Plugins {
+		if plugin.Network.Type == "gke" {
+			gkePluginFound = true
+			ipamConfig, _, err = allocator.LoadIPAMConfig(plugin.Bytes, "")
+			if err != nil {
+				glog.Errorf("Error loading IPAM config: %v", err)
+				return err
+			}
+			if len(ipamConfig.Ranges) == 0 {
+				return fmt.Errorf("no subnets assigned in the IPAM config")
+			}
+			break
+		}
+	}
+	if !gkePluginFound {
+		return fmt.Errorf("gke plugin not found in network config")
+	}
+
+	for _, ranges := range ipamConfig.Ranges {
+		for _, ipRange := range ranges {
+			totalIP, err := c.countIPsFromRange(ipRange.Subnet)
+			if err != nil {
+				glog.Errorf("Error calculating total IPs for subnet %s: %v", ipRange.Subnet.IP.String(), err)
+				return err
+			}
+			c.updateAssignedIPs(ipRange.Subnet, totalIP)
+		}
+	}
 	return nil
 }
 
 func (c *podIPMetricsCollector) Update(ch chan<- prometheus.Metric) error {
-	if !podIPMetricsWatcherSetup {
+	if !podIPMetricsWatcherIsIntialized {
 		if err := c.setupDirectoryWatcher(gkePodNetworkDir); err != nil {
 			glog.Errorf("setupDirectoryWatcher returned error: %v", err)
 			return nil
 		}
+		//calculate assigned IP count
+		if err := c.calculateAssignedIP(gkeCNIConfigFile); err != nil {
+			glog.Errorf("calculateAssignedIP returned error: %v", err)
+		}
 	}
+
 	ch <- prometheus.MustNewConstMetric(usedIPv4AddrCountDesc, prometheus.GaugeValue, float64(c.usedIPv4AddrCount))
 	ch <- prometheus.MustNewConstMetric(usedIPv6AddrCountDesc, prometheus.GaugeValue, float64(c.usedIPv6AddrCount))
 	ch <- prometheus.MustNewConstMetric(dualStackCountDesc, prometheus.GaugeValue, float64(c.dualStackCount))
 	ch <- prometheus.MustNewConstMetric(dualStackErrorCountDesc, prometheus.GaugeValue, float64(c.dualStackErrorCount))
 	ch <- prometheus.MustNewConstMetric(duplicateIPCountDesc, prometheus.GaugeValue, float64(c.duplicateIPCount))
-
 	ch <- prometheus.MustNewConstMetric(ipReuseMinDesc, prometheus.GaugeValue, float64(c.reuseIPs.min))
 	ch <- prometheus.MustNewConstMetric(ipReuseAvgDesc, prometheus.GaugeValue, c.reuseIPs.sum/float64(c.reuseIPs.count))
 	ch <- prometheus.MustNewConstHistogram(ipReuseHistogramDesc, c.reuseIPs.count, c.reuseIPs.sum, c.reuseIPs.buckets)
-
+	ch <- prometheus.MustNewConstMetric(assignedIPv4AddrCountDesc, prometheus.GaugeValue, float64(c.assignedIPv4AddrCount))
+	ch <- prometheus.MustNewConstMetric(assignedIPv6AddrCountDesc, prometheus.GaugeValue, float64(c.assignedIPv6AddrCount))
```

Comment:
Can we only export these two metrics when `calculateAssignedIP()` succeeded? so we are not exposing 0 value when it failed e.g. the calico case.

Reply:
Done.